### DISCUSSION TITLE
Creating Cursor from Path / vector graphic

### DISCUSSION THREAD

**Discussion by stevemonaco** [AvaloniaUI/Avalonia#13334] [Q&A] [answered]
I'm trying to create a `Cursor` based upon a vector graphic (`Path`, obtained from MahApps.Metro.IconPacks). My current approach is to render the `Path` onto a `RenderTargetBitmap` which is used to create the `Cursor` with. I assume I'm missing something fundamental about scaling or `RenderTargetBitmap`.

The particular vector artwork is quite dimensionally-large coming from IconPacks and requires a ~512x512 surface. I've made a [repro](https://github.com/stevemonaco/RenderTargetBitmapWithPath) with a few scenarios:

1. OriginalPathCursor - Rendered on a 512x512 surface. Works, albeit enormous
2. OriginalPathBarelyShowingCursor - Rendered on a 180x150 surface. Works, but is clipped and barely showing.
3. OriginalPathTooSmallCursor - Rendered on a 32x32 surface. White rectangle. Seems to happen when no content is rendered in bounds.
4. ManuallyScaledPathCursor - Rescaled in Inkscape and rendered on a 32x32 surface. Works, but manually scaling it not desired.
5. PathIconPathCursor - Attempted to force layout (measure+arrange) and rendered on a 32x32 surface. White rectangle.

I save each scenario to a PNG file which shows the same result as the cursor, so I'm sure the problem is in my rendering and not creating the cursor. Most work in the repro was done in the VM for easy testing and I'm aware it's not MVVM-pure.

[Repro](https://github.com/stevemonaco/RenderTargetBitmapWithPath)

Avalonia 11.0.5 / Windows 10

![cursors](https://github.com/AvaloniaUI/Avalonia/assets/1782158/bc02495a-e0ea-48ce-92d0-86308ab6c74a)

**Comment by stevemonaco** [comment, marked as answer]
This seems to work with a `Viewbox` containing a `Path`. Listing the relevant bits here since I'll probably remove the repro eventually. I am still open to other approaches that may be a bit smarter in centering.

Factory:
```cs
public static class VisualCursorFactory
{
    public static Cursor Create(Visual visual, PixelSize size, PixelPoint hotSpot)
    {
        using var rtb = new RenderTargetBitmap(size, new(96, 96));
        rtb.Render(visual);
        return new Cursor(rtb, hotSpot);
    }
}
```

Usage:
```cs
var originalPath = new Path
{
    Data = StreamGeometry.Parse("M256 124l68 68l30 -30l-98 -98l-98 98l30 30zM256 388l-68 -68l-30 30l98 98l98 -98l-30 -30z"),
    StrokeThickness = 2,
    Fill = Brushes.Orange,
    Stroke = Brushes.Red,
};

int length = 32;
var vb = new Viewbox
{
    Width = length,
    Height = length,
    Child = originalPath,
};

vb.Measure(new(length, length));
vb.Arrange(new Avalonia.Rect(0, 0, length, length));

ViewboxPathCursor = VisualCursorFactory.Create(vb, new Avalonia.PixelSize(length, length), new Avalonia.PixelPoint(length / 2, length / 2));
```

XAML:
```xml
<TextBlock Cursor="{Binding ViewboxPathCursor}" Text="Via Viewbox" />
```

**Reply by timunie** [reply]
LayoutTransformControl instead of ViewBox should also work
